node/pkg/watchers/evm/connectors/poller.go

```diff
@@ -53,186 +46,169 @@ func NewBlockPollConnector(ctx context.Context, baseConnector Connector, finaliz
+func (b *BlockPollConnector) pollBlock(ctx context.Context, logger *zap.Logger, prevLatest *NewBlock, prevFinalized *NewBlock) (newLatest *NewBlock, newFinalized *NewBlock, err error) {
+	newLatest, err = getBlockByTag(ctx, logger, b.Connector, "latest", Latest)
 	if err != nil {
-		logger.Error("failed to look up latest block",
-			zap.Uint64("lastSeenBlock", lastBlock.Number.Uint64()), zap.Error(err))
-		return lastPublishedBlock, fmt.Errorf("failed to look up latest block: %w", err)
+		err = fmt.Errorf("failed to get latest block: %w", err)
+		newLatest = prevLatest
+		newFinalized = prevFinalized
+		return
 	}
-	for {
-		if lastPublishedBlock.Number.Cmp(latestBlock.Number) >= 0 {
-			// We have to wait for a new block to become available
-			return
-		}
 
-		// Try to fetch the next block between lastBlock and latestBlock
-		nextBlockNumber := new(big.Int).Add(lastPublishedBlock.Number, big.NewInt(1))
-		block, err := b.getBlockWithTimeout(ctx, logger, nextBlockNumber, safe)
-		if err != nil {
-			logger.Error("failed to fetch next block",
-				zap.Uint64("block", nextBlockNumber.Uint64()), zap.Error(err))
-			return lastPublishedBlock, fmt.Errorf("failed to fetch next block (%d): %w", nextBlockNumber.Uint64(), err)
-		}
-
-		finalized, err := b.isBlockFinalizedWithTimeout(ctx, block)
-		if err != nil {
-			logger.Error("failed to check block finalization",
-				zap.Uint64("block", block.Number.Uint64()), zap.Error(err))
-			return lastPublishedBlock, fmt.Errorf("failed to check block finalization (%d): %w", block.Number.Uint64(), err)
-		}
+	// First see if there might be some newly finalized ones to publish
+	var block *NewBlock
+	if newLatest.Number.Cmp(prevLatest.Number) > 0 {
+		// If there is a gap between prev and new, we have to look up the transaction hashes for the missing ones. Do that in batches.
+		newBlockNum := newLatest.Number.Uint64()
+		for blockNum := prevLatest.Number.Uint64() + 1; blockNum < newBlockNum; blockNum++ {
+			block, err = getBlockByNumberUint64(ctx, logger, b.Connector, blockNum, Latest)
+			if err != nil {
+				err = fmt.Errorf("failed to get gap block: %w", err)
+				newLatest = prevLatest
+				newFinalized = prevFinalized
+				return
+			}
 
-		if !finalized {
-			break
+			b.blockFeed.Send(block)
 		}
 
-		b.blockFeed.Send(block)
-		lastPublishedBlock = block
+		b.blockFeed.Send(newLatest)
+	} else if newLatest.Number.Cmp(prevLatest.Number) < 0 {
+		logger.Error("latest block number went backwards, ignoring it", zap.Any("newLatest", newLatest), zap.Any("prevLatest", prevLatest))
+		newLatest = prevLatest
 	}
 
-	return
-}
+	newFinalized = prevFinalized
+	if newLatest.Number.Cmp(prevFinalized.Number) > 0 {
+		var finalized bool
+		// If there is a gap between prev and new, we have to look up the transaction hashes for the missing ones. Do that in batches.
+		newBlockNum := newLatest.Number.Uint64()
```

Comment:
This looks almost the same as lines 134 - 149.  Why are we doing it again?

Reply:
Above is for latest. This is for finalized. I added comments.